src/octoprint/plugins/gcodeviewer/static/js/viewer/reader.js

```diff
@@ -77,34 +80,116 @@ GCODE.gCodeReader = (function () {
         return tmpModel;
     };
 
-    var prepareLinesIndex = function (m) {
-        percentageTree = undefined;
-
-        for (var l = 0; l < m.length; l++) {
-            if (m[l] === undefined) continue;
-            for (var i = 0; i < m[l].length; i++) {
-                var percentage = m[l][i].percentage;
-                var value = {layer: l, cmd: i};
-                if (!percentageTree) {
-                    percentageTree = new AVLTree({key: percentage, value: value}, "key");
+    var searchInPercentageTree = function (key) {
+        function searchInLayers(lower, upper, key) {
+            while (lower < upper) {
+                var middle = Math.floor((lower + upper) / 2);
+
+                if (
+                    rendererModel[middle][0].percentage <= key &&
+                    (!rendererModel[middle + 1] ||
+                        rendererModel[middle + 1][0].percentage > key)
+                )
+                    return middle;
+
+                if (rendererModel[middle][0].percentage > key) {
+                    upper = middle - 1;
                 } else {
-                    percentageTree.add({key: percentage, value: value});
+                    lower = middle + 1;
                 }
             }
+
+            return lower;
         }
-    };
 
-    var searchInPercentageTree = function (key) {
-        if (percentageTree === undefined) {
-            return undefined;
+        function searchInCmds(layer, lower, upper, key) {
+            while (lower < upper) {
+                var middle = Math.floor((lower + upper) / 2);
+
+                if (rendererModel[layer][middle].percentage == key) return middle;
+
+                if (rendererModel[layer][middle].percentage > key) {
+                    upper = middle - 1;
+                } else {
+                    lower = middle + 1;
+                }
+            }
+            return lower;
         }
 
-        var elements = percentageTree.findBest(key);
-        if (elements.length === 0) {
-            return undefined;
+        if (rendererModel === undefined) return undefined;
+
+        // this happens when the print is stopped.
+        // just return last position to keep the last
+        // position on screen.
+        if (key == null) return {layer: cacheLastLayer, cmd: cacheLastCmd};
+
+        var bestLayer = undefined;
+        var bestCmd = undefined;
+
+        // check if we are within cacheLookAhead distance of our last position
+        if (cacheLastLayer !== undefined) {
+            if (
+                rendererModel[cacheLastLayer][0].percentage <= key &&
+                (!rendererModel[cacheLastLayer + 1] ||
+                    rendererModel[cacheLastLayer + 1][0].percentage > key)
+            ) {
+                bestLayer = cacheLastLayer;
+
+                var upper =
+                    (cacheLastCmd + cacheLookAhead) % rendererModel[bestLayer].length;
```

Comment:
Am I missing something here or is this really just a confusing way to say

```suggestion
                var upper =
                    Math.min(cacheLastCmd + cacheLookAhead, rendererModel[bestLayer].length - 1);
```